### ISSUE TITLE
Visual Studio Instrumentation profiling integration?

### ISSUE THREAD

**Issue by Rseding91** [fastbuild/fastbuild#1034]
We currently use FastBuild for our codebase and I was wondering if and or how to properly configure our fastbuild config to get Visual Studios "instrumentation" profiling working with it. When I attempt to use the instrumentation profiling option it simply says none of the projects were built with /PROFILE. I was unsuccessful in attempting to add that anywhere.

I'm not sure if this has ever been looked into (I can't find anything about it online). So I don't know if I'd need to modify out config every time I want to use this profiling mode or if fastbuild can detect when it's invoked and add it itself (would be ideal).

**Comment by Rseding91** [comment]
Right, but when I add that, it still does not work and produces the exact same error - which is what led me here. No matter where and how I add the linker option ' /PROFILE' visual studio 2022 continues to report "Linker is not set to /PROFILE" for all projects.

![image](https://github.com/fastbuild/fastbuild/assets/2196754/40222a3a-2e1b-42cc-8514-767af142dbfc)

**Comment by ffulin** [comment]
Thanks for the extra info. I'm not overly familiar with this specific profiler (I've not had much success getting useful results out of any of the built-in MSVC profilers so I haven't used them in years). In my previous response I was assuming you were using the command line VSInstr.exe.

With VSInstr.exe on the command line, I had:
```
VSInstr.exe c:\p4\depot\tmp\x64-Release\Core\CoreTest\CoreTest.exe /OUTPUTPATH:c:\p4\depot\tmp\x64-Release\Core\CoreTest\I\
...
Unable to obtain debug information. Link with the /PROFILE linker switch.
```
and then adding /PROFILE got me:
```
VSInstr.exe c:\p4\depot\tmp\x64-Release\Core\CoreTest\CoreTest.exe /OUTPUTPATH:c:\p4\depot\tmp\x64-Release\Core\CoreTest\I\
...
Analyzing component...
Instrumenting component...
#################### (100.00 %%)
Successfully instrumented file c:\p4\depot\tmp\x64-Release\Core\CoreTest\CoreTest.exe.
```

Running this supposedly instrumented exe didn't really do anything obviously useful though.

Using the IDE, CPU profiling works find out of the box, but I was able to repro the issue you screenshotted when choosing "Instrumentation" instead.

I spent a bit more time looking into this to see if I could coax VS into working.

I found that this seemed to be present in the those working projects:
```
    <Link>
      <Profile>true</Profile>
    </Link>
```
I tried various ways of hand editing the FASTBuild generated project and adding that in various places just to see if it's possible to get a working project, but nothing I could do would satisfy the IDE.

I also tried pointing the project properties at the exe i "instrumented" from the command line above, but that didn't seem to do anything either.

In the end I unfortunately wasn't able to get things working from the UI in Visual Studio for either a FASTBuild generated project, or a vanilla Makefile project (made using the Visual Studio wizard). I could only get things working for a non-makefile Visual Studio project.

From here, potential avenues to try pursuing:
**a) open a ticket with MS support.** Perhaps there's a way to make it work nicely with makefile projects we're missing
If there's something we can do to generate a working project, FASTBuild could almost certainly be easily updated to do that.
It's not uncommon for MS to treat makefile projects as a bit of a second-class citizen unfortunately, but maybe we missed something or perhaps there is something they can fix. (They have made improvements to makefile projects over the years)
**b) See if the VSDiagnotsics.exe can be used** from the command line (it lives next to VSInstr.exe) Perhaps it can launch the instrumented exe created by VSInstr.exe in a useful way?
(https://learn.microsoft.com/en-us/visualstudio/profiling/profile-apps-from-command-line?view=vs-2022)
**c) Create a fake non-makefile MSVC project** - point it to our already built (and "instrumented" if necessary) executable instead of what it would actually build and see if that can be made to work.

**Comment by Rseding91** [comment]
I'll try option A. I was recently talking with a member of the visual studio profiling team about this so maybe they can assist.

**Comment by ffulin** [comment]
Great! If you can, let me know the outcome and whether something can be improved no the FASTBuild side.

**Comment by Rseding91** [comment]
I never got anything moving with this, and testing manually adding /profile to the build process results in VS getting stuck with our project. For now, it seems instrumentation is not viable once a project goes above some size/complexity.

**Comment by ffulin** [comment]
Thanks for the update.

On a related note, I can recommend Superluminal as a great sampling based profiler. That also includes command line tooling which can be useful.
